server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java

```diff
@@ -61,11 +62,10 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
                 )
             );
 
-        synchronized (tasksPerBatchingKey) {
-            LinkedHashSet<BatchedTask> existingTasks = tasksPerBatchingKey.computeIfAbsent(
-                firstTask.batchingKey,
-                k -> new LinkedHashSet<>(tasks.size())
-            );
+        tasksPerBatchingKey.compute(firstTask.batchingKey, (k, existingTasks) -> {
+            if (existingTasks == null) {
+                return new LinkedHashSet<>(tasks);
+            }
             for (BatchedTask existing : existingTasks) {
                 // check that there won't be two tasks with the same identity for the same batching key
```

Comment:
FWIW I don't think I've ever seen this check fail, maybe we could make it an assertion?

Reply:
I'm all for it. I think we can only get here via a bug and I've also never seen this happen outside of bugs during experimenting with stuff. I'll make it an assertion :)